feat: update to @open-wc/scoped-element not needing a polyfill (#1647)
ing-bank/lion: docs/blog/lion-without-polyfills.md

---
title: Lion without poylfills
published: true
description: Lion has been a long user of the scoped registry - always requiring a polyfill - but no more. Load the polyfill only if you need it.
date: 2022-04-04
tags: [javascript, polyfills]
cover_image: /blog/images/introducing-lions-website-cover-image.jpg
---

> From the start we have been focused on making it very explicit which sub component are used within a given custom element.

Author: added

> This also scoped these sub components to the current shadow root.

Author: good idea - added 👍

---
title: Lion without poylfills
published: true
description: Lion has been a long user of the scoped registry - always requiring a polyfill - but no more. Load the polyfill only if you need it.
date: 2022-04-04
tags: [javascript, polyfills]
cover_image: /blog/images/introducing-lions-website-cover-image.jpg
---

From the start we have been focused on making it very explicit which sub component are used within a given custom element.
This also scoped these sub components to the current shadow root.

> This means that you can use a lion components like `lion-listbox` and never have to worry about if the internally used components clash with others you are already using.

Reviewer: component*

---
title: Lion without poylfills
published: true
description: Lion has been a long user of the scoped registry - always requiring a polyfill - but no more. Load the polyfill only if you need it.
date: 2022-04-04
tags: [javascript, polyfills]
cover_image: /blog/images/introducing-lions-website-cover-image.jpg
---

> From the start we have been focused on making it very explicit which sub component are used within a given custom element.

Reviewer: components*

> From the start we have been focused on making it very explicit which sub component are used within a given custom element.

Reviewer: ```suggestion
From the start we have been focused on making it very explicit which sub components are used within a given custom element.
```

> From the start we have been focused on making it very explicit which sub component are used within a given custom element.

Author: ```suggestion
From the start, we have been focused on making it very explicit which sub-components are used within a given custom element.
```

> This is commonly revered to as [component composition](https://lit.dev/docs/composition/component-composition/).

Reviewer: referred*

> This is commonly revered to as [component composition](https://lit.dev/docs/composition/component-composition/).

Reviewer: ```suggestion
This is commonly referred to as [component composition](https://lit.dev/docs/composition/component-composition/).
```

> This is commonly revered to as [component composition](https://lit.dev/docs/composition/component-composition/).

Author: ```suggestion
This is commonly referred to as [component composition](https://lit.dev/docs/composition/component-composition/).
```

On top of that we scoped these [sub components](https://open-wc.org/docs/development/scoped-elements/) to the [current shadow root](https://github.com/WICG/webcomponents/blob/gh-pages/proposals/Scoped-Custom-Element-Registries.md).

> This means that you can use a lion component like `lion-listbox` and never have to worry about if the internally used components clash with others you are already using.

Reviewer: Maybe we should mention here that `lion-listbox` uses LionOptions internally as 'lion-options'/

Even better: give a more hypothetical example, since most of our internal compositions are workarounds for the to be implemented [a11y spec](https://wicg.github.io/aom/notification-api.html).

So the most simple and understandable use case for a consumer would be shadow dom usage like `MyCard/my-card` that uses `MyCardHeader/my-card-header`

Author: the example below is for Listbox? 🤔

not clear enough? or worth repeating here?

Reviewer: I would go with a purely hypothetical one consumers could immediately relate to without knowing our codebase and without the complexity of needing to be rendered to light dom (apart from the fact that inside our codebase it's implemented way different).
(MyLionCard and MyLionCardHeader would have a really low 'cognitive load')

## How does it work?

> 1. Within lion class files we only import other class files

Reviewer: ```suggestion
1. Within Lion class files we only import other class files (e.g. in class `MyLionCard` we use `MyLionCardHeader` via composition)
```

> 2. We define them as `scopedElements` and let the mixin handle the rest

Reviewer: ```suggestion
2. We define them as `scopedElements` (`my-lion-card-header: MyLionCardHeader `) and let the ScopedElmentsMixin handle the rest
```

> To clarify within lion class files we never import files that run `customElement.define`

Reviewer: ```suggestion
To clarify: within lion class files we never import files that run `customElement.define`
```

> To clarify within lion class files we never import files that run `customElement.define`

Author: ```suggestion
To clarify, within lion class files we never import files that run `customElement.define`
```

```js
import { LitElement, html } from 'lit';
import { ScopedElementsMixin } from '@open-wc/scoped-elements';
import { LionOptions } from './LionOptions.js';

export class LionListbox extends ScopedElementsMixin(LitElement) {
  static scopedElements = {
    'lion-options': LionOptions,
  };

  render() {
    return html`
      <div>
        <!-- example code - not the real implementation -->
        <lion-option></lion-option>
      </div>
    `;
  }
}
```

> ## What is wrong with it?

Reviewer: The title seems to suggest smth is wrong with the above implementation, whereas it was the prev version of ScopedElementsMixin was wrong.

Maybe make it "## Known challenges in previous releases"

The code above totally makes sense - however we always assumed that a scoped registry will be available.
Which was somewhat of a valid assumption as all our components are using the `ScopedElementsMixin` and it in turn loads a polyfill for the scoped registry.

> We however over time we got multiple feedbacks that lion components "break the app as soon as you load them".

Reviewer: ```suggestion
We however over time we got feedback from multiple consumers that lion components "break the app as soon as you load them".
```

> The reasons is/was that not everyone is always using `ScopedElementsMixin` or in full control of the app (or it's load order).

Reviewer: ```suggestion
The reasons is/was that not everyone is always using `ScopedElementsMixin` or in full control of the app (or its load order).
```

To quote the release notes of the latest version of `ScopedElementsMixin`:

> ScopedElementsMixin 2.x tried to be as convenient as possible by automatically loading the scoped custom elements registry polyfill.
> This however led to a fatal error whenever you registered any component before ScopedElementsMixin was used.

And this was the case.

## How do we fix it?

With the latest release of lion we now updated to the latest version of `ScopedElementsMixin` which means lion now works in all apps as long as there is no need for actual scoping.

> If you extends lion components and you imperatively create scoped custom elements then you should now use a helper function that will work in scoped and unscoped cases.

Reviewer: ```suggestion
If you extend lion components and you imperatively create scoped custom elements, you should now use a helper function that will work in scoped and unscoped cases.
```

```diff
-  const myButton = this.shadowRoot.createElement('my-button');
+  const myButton = this.createScopedElement('my-button');
```

## How to get scoping

> You need scoping if you want to

Reviewer: ```suggestion
You need scoping if you want to:
```

- use 2 major versions of a web component (e.g. in an SPA pageA uses 1.x and pageB uses 2.x of color-picker)

> - or you want to use the same tag name with different implementations (use tag color-picker from foo here and from bar here)

Reviewer: ```suggestion
- use the same tag name with different implementations (use tag color-picker from foo here and from bar here)
```

If you need scoping you need to install and load the polyfill first thing in your HTML.

```bash
npm i @webcomponents/scoped-custom-element-registry
```

It could look something like this:

```html
<script src="/node_modules/@webcomponents/scoped-custom-element-registry/scoped-custom-element-registry.min.js"></script>
```

or if you have an SPA you can load it at the top of your app shell code

```js
import '@webcomponents/scoped-custom-element-registry';
```

> 

Reviewer: In this scoping section, we should emphasize this is desired behavior and that as soon as scoped registries are supported by the platform, the behavior that you get with loading, will be the default.

Reviewer: Then I would expect a section here that refers to the title... (w/o polyfills)
Explain the need for it (perf in small apps) and how it is backw. compatible with enabling the registry (either via browser or polyfill)?
